A55-xds-stateful-session-affinity.md

```diff
@@ -0,0 +1,480 @@
+configured LB policy. When gRPC receives the response for that first RPC from
+B1, it attaches a configured cookie to the response before delivering the
+response to the client application. The cookie encodes some identifying
+information about B1 (for example its IP-address). The client application
+saves the cookie and uses that cookie in all future RPCs of that session
+(e.g. by using a separate cookie jar for each session). gRPC uses the
+cookie and gets the encoded backend information to route the RPC to the
+same backend B1 if it is still reachable.
+
+With stateful session affinity, session draining of backends is an important
+requirement. In this use-case, the user wants to stop or restart service
+backends (let's say to optimize resources or to perform upgrades). However a
+terminating backend first needs to "drain" all assigned sessions before
+exiting. In this state - let's call it `DRAINING` - the terminating backend is
+only sent RPCs of its existing assigned sessions and no new traffic is sent.
+Once all assigned sessions have finished (or a timeout triggers), the backend
+application exits at which point the backend is now removed from the endpoints
+list (EDS). This design includes support for session draining.
+
+### Related Proposals: 
+
+* [A27: xDS-Based Global Load Balancing][A27]
+* [A42: xDS Ring Hash LB Policy][A42]
+
+## Proposal
+
+The goal of this design is to make it consistent with the Envoy implementation
+so it will interoperate with it. For example, one can create an xDS
+configuration for Envoy and the exact same configuration will work with
+proxyless gRPC. Similarly, an application using some cookie implementation
+that works with Envoy will continue to work with proxyless gRPC without any
+modifications.
+
+The design involves following 5 parts.
+
+### `XdsClient` to Include Endpoint Health Status
+
+The `EdsUpdate` struct from the `XdsClient` will be enhanced to include the
+[endpoint health status][eds-health-status] for each endpoint. `XdsClient`
+will also be modified to include endpoints in `DRAINING` state in the
+`EdsUpdate` reported to the watchers.
+
+### Listener and Route Configuration through xDS
+
+The feature is enabled and configured using the
+[cookie based stateful session extension][cookie-ext] which
+is an [http-filter][] in the `HttpConnectionManager` for the target service.
+A sample configuration is as follows:
+
+```textproto
+{ // HttpConnectionManager
+  http_filters: [
+    {
+      name: "envoy.filters.http.stateful_session"
+      typed_config: {
+        "@type": "type.googleapis.com/envoy.extensions.filters.http.stateful_session.v3.StatefulSession"
+        session_state: {
+          name: "envoy.http.stateful_session.cookie"
+          typed_config: {
+            "@type": "type.googleapis.com/envoy.extensions.http.stateful_session.cookie.v3.CookieBasedSessionState"
+            cookie: {
+              name: "global-session-cookie"
```

Comment:
I don't think 'global-session-cookie' is right - with the current design it should be "service-session-cookie".

Also I would remove the path, or at least the method - it's not what we'll implement, and not clear if having different methods go to different backends is desirable.

Reply:
This is just an example as to what the name of the cookie might be.  gRPC doesn't care about the name; we'll use whatever name the control plane tells us to use.

The existing xDS API that we're implementing here [has a path field](https://github.com/envoyproxy/envoy/blob/dcd329a2e95b54f754b17aceca3f72724294b502/api/envoy/type/http/v3/cookie.proto#L30), so we're going to support it.  Istio can choose whether or not it's going to support setting it, but the client needs to behave the right way even if it's talking to a control plane other than Istio.